src/Text/Pretty/Simple/Internal/OutputPrinter.hs

```diff
@@ -176,82 +185,130 @@ hCheckTTY h options = liftIO $ conv <$> tty
+prettyExpr opts = \case
+  Brackets xss -> list "[" "]" xss
+  Braces xss -> list "{" "}" xss
+  Parens xss -> list "(" ")" xss
+  StringLit s -> join enclose (annotate Quote "\"") $ annotate String $ pretty s
+  CharLit s -> join enclose (annotate Quote "'") $ annotate String $ pretty s
+  Other s -> pretty s
+  NumberLit n -> annotate Num $ pretty n
   where
-    process :: OutputOptions -> String -> String
-    process opts = case outputOptionsStringStyle opts of
-      Literal -> id
-      EscapeNonPrintable ->
-        indentSubsequentLinesWith spaces . escapeNonPrintable . readStr
-      DoNotEscapeNonPrintable -> indentSubsequentLinesWith spaces . readStr
-      where
-        spaces :: String
-        spaces = replicate (indentSpaces + 2) ' '
-
-        indentSpaces :: Int
-        indentSpaces =  outputOptionsIndentAmount opts
-
-        readStr :: String -> String
-        readStr s = fromMaybe s . readMaybe $ '"':s ++ "\""
-renderOutput (Output _ (OutputCharLit string)) = do
-  sequenceFold
-    [ useColorQuote
-    , pure "'"
-    , useColorReset
-    , useColorString
-    , pure (fromString string)
-    , useColorReset
-    , useColorQuote
-    , pure "'"
-    , useColorReset
-    ]
+    list :: Doc Annotation -> Doc Annotation -> CommaSeparated [Expr]
+      -> Doc Annotation
+    list open close (CommaSeparated xss) =
+      enclose (annotate Open open) (annotate Close close) $ case xss of
+        [] -> mempty
+        [xs] | all isSimple xs ->
+          space <> hsep (map (prettyExpr opts) xs) <> space
+        _ -> concatWith lineAndCommaSep (map (prettyExprs opts) xss) <> line
+    lineAndCommaSep x y = x <> line <> annotate Comma "," <> y
+
+-- | Determine whether this expression should be displayed on a single line.
+isSimple :: Expr -> Bool
+isSimple = \case
+  (getList -> Just [[e]]) -> isSimple e
+  (getList -> Just (_:_)) -> False
+  _ -> True
+  where
+    getList = \case
+      Brackets (CommaSeparated xs) -> Just xs
+      Braces (CommaSeparated xs) -> Just xs
+      Parens (CommaSeparated xs) -> Just xs
+      _ -> Nothing
+
+-- | Traverse the stream, using a 'Tape' to keep track of the current color.
+annotateAnsi :: OutputOptions -> SimpleDocStream Annotation
+  -> SimpleDocStream AnsiStyle
+annotateAnsi opts ds = case outputOptionsColorOptions opts of
+  Nothing -> unAnnotateS ds
+  Just ColorOptions {..} -> evalState (traverse style ds) initialTape
+    where
+      style :: MonadState (Tape AnsiStyle) m => Annotation -> m AnsiStyle
+      style = \case
+        Open -> modify moveR *> gets tapeHead
+        Close -> gets tapeHead <* modify moveL
+        Comma -> gets tapeHead
+        Quote -> pure colorQuote
+        String -> pure colorString
+        Num -> pure colorNum
+      initialTape = Tape
+        { tapeLeft = streamRepeat colorError
+        , tapeHead = colorError
+        , tapeRight = streamCycle $ fromMaybe (pure mempty)
+            $ nonEmpty $ colorRainbowParens
+        }
```

Comment:
This is really ingenious!

I was wondering whether this would alternatively have been a good fit for [`SimpleDocTree`](http://hackage.haskell.org/package/prettyprinter-1.6.2/docs/Data-Text-Prettyprint-Doc-Render-Util-SimpleDocTree.html), but I think it might have been tricky to colour the parentheses and the contents differently…

Reply:
Thanks! I was quite happy with how elegantly this bit turned out.

I considered `SimpleDocTree`, but couldn't quite see how to make it work.